File: AquaQ/Challenge_01/AquaQ_01.py
"""AquaQ Puzzles - Puzzle 1
Solution Started: September 62025,
Puzzle Link: https://challenges.aquaq.co.uk/challenge/1
Solution by: Abbas Moosajee
Brief: [Rose by any other name]"""

#!/usr/bin/env python3
from pathlib import Path

# Load input file
input_file = "Challenge_01_input.txt"
input_path = Path(__file__).parent / input_file

with input_path.open("r", encoding="utf-8") as f:
    data = f.read().splitlines()

HEXADECIMAL_CHARS = "0123456789abcdef"

def to_hex_string(s: str) -> str:
    """Replace non-hex chars with '0'."""
    return "".join(c if c in HEXADECIMAL_CHARS else "0" for c in s)

def pad_to_next_multiple_of_3(s: str, pad_char="X") -> str:
    """Pad string length up to nearest multiple of 3."""
    pad_len = (3 - len(s) % 3) % 3
    return s + pad_char * pad_len

def process_string(data: str) -> str:
    hex_chars = to_hex_string(data)
    padded = pad_to_next_multiple_of_3(hex_chars)
    split_size = len(padded) // 3

    # take first 2 chars of each section
    sections = [padded[i:i+split_size] for i in range(0, len(padded), split_size)]
    return "".join(sec[:2] for sec in sections)

# Example
final_string = process_string(data[0])
print("Challenge 01:", final_string)

File: AquaQ/src/bin/AquaQ_01.rs
/* AquaQ Puzzles - Puzzle 1
Solution Started: September 72025,
Puzzle Link: https://challenges.aquaq.co.uk/challenge/1
Solution by: Abbas Moosajee
Brief: [Rose by any other name] */

use std::{fs, env, error::Error};
use std::collections::HashSet;

const HEXADECIMAL_CHARS: &str = "0123456789abcdef";

/// Replace non-hex chars with '0'
fn to_hex_string(s: &str) -> String {
    let hexset: HashSet<char> = HEXADECIMAL_CHARS.chars().collect();
    s.chars()
        .map(|c| if hexset.contains(&c) { c } else { '0' })
        .collect()
}

/// Pad string length to nearest multiple of 3
fn pad_to_next_multiple_of_3(s: &str, pad_char: char) -> String {
    let pad_len = (3 - (s.len() % 3)) % 3;
    let mut result = String::from(s);
    for _ in 0..pad_len {
        result.push(pad_char);
    }
    result
}

/// Process string: sanitize, pad, split, and join
fn process_string(data: &str) -> String {
    let hex_chars = to_hex_string(data);
    let padded = pad_to_next_multiple_of_3(&hex_chars, 'X');
    let split_size = padded.len() / 3;

    padded
        .as_bytes()
        .chunks(split_size)
        .map(|chunk| {
            let part: String = chunk.iter().map(|&b| b as char).collect();
            part.chars().take(2).collect::<String>()
        })
        .collect::<Vec<String>>()
        .join("")
}

fn main() -> Result<(), Box<dyn Error>> {

    // Get the first command-line argument, or default to "Lang06_input.txt"
    let args: Vec<String> = env::args().collect();
    let file_path = if args.len() > 1 {
        &args[1]  // Use the provided file path
    } else {
        "../../inputs/challenge_01.txt"  // Default file
    };

    let contents = fs::read_to_string(file_path)?;

    let data: Vec<&str> = contents
        .lines()
        .map(|line| line.trim())
        .filter(|line| !line.is_empty())
        .collect();

    let final_string = process_string(data[0]);
    println!("Challenge 01: {}", final_string);

    Ok(())
}
